Add Monitoring Multicluster Istio with Prometheus doc (#7040)
istio/istio.io: content/en/docs/ops/configuration/telemetry/monitoring-multicluster-prometheus/index.md

[hierarchical federation](https://prometheus.io/docs/prometheus/latest/federation/#hierarchical-federation). 
Cluster-local Istio-deployed Prometheus instances will act as initial collectors which then federate up to a production 
mesh-wide Prometheus instance. That mesh-wide Prometheus can either live outside of all of the clusters from the mesh 
(aka external), or in one of the clusters within the mesh.

## Multicluster Istio Setup
There are a couple of basic [multi-cluster deployment](/docs/ops/deployment/deployment-models/#multiple-clusters) 
architectures supported by Istio. For the purposes of this guide, any of those approaches will work, with the following 
caveat:

**Ensure that a cluster-local Istio Prometheus instance is installed in each cluster.**

Individual Istio deployments of Prometheus in each cluster will be used to form the basis of cross-cluster monitoring by
way of federation to either an external production-ready Prometheus or a production-ready instance of Prometheus 
running on one of the clusters.

For multi-cluster deployments that use the remote profile, this will require adding the following to the istioctl 

> manifest command:

Reviewer: ```suggestion
manifest command:

```

---
title: Monitoring Multicluster Istio with Prometheus
description: Configure Prometheus to monitor multicluster Istio.
weight: 10
aliases:
  - /help/ops/telemetry/monitoring-multicluster-prometheus
  - /docs/ops/telemetry/monitoring-multicluster-prometheus
---

## Overview

> This is meant to provide operational guidance on how to configure monitoring of Istio meshes constituted by two

Reviewer: ```suggestion
This is meant to provide operational guidance on how to configure monitoring of Istio meshes comprised of two
```

Author: done

> This is meant to provide operational guidance on how to configure monitoring of Istio meshes comprised of two
> or more individual Kubernetes clusters. It is not meant to establish the *only* possible path forward, but rather
> to demonstrate a workable approach to multicluster telemetry with Prometheus.

Reviewer 2: ```suggestion
This guide explains how to configure multicluster telemetry with Prometheus for an Istio mesh that consists of two or more Kubernetes clusters.
```

Our recommendation for multicluster monitoring of Istio with Prometheus is built upon the foundation of Prometheus
[hierarchical federation](https://prometheus.io/docs/prometheus/latest/federation/#hierarchical-federation).

> Cluster-local Istio-deployed Prometheus instances will act as initial collectors which then federate up to a production

Author: also, I think `act as` should be changed to `acting as`?

> mesh-wide Prometheus instance. That mesh-wide Prometheus can either live outside of all of the clusters from the mesh

Reviewer: ```suggestion
mesh-wide Prometheus instance. That mesh-wide Prometheus can either live outside of the mesh
```

Author: done

> Our recommendation for multicluster monitoring of Istio with Prometheus is built upon the foundation of Prometheus
> [hierarchical federation](https://prometheus.io/docs/prometheus/latest/federation/#hierarchical-federation).
> Prometheus instances that are deployed locally to each cluster by Istio act as initial collectors that then federate up to a production
> mesh-wide Prometheus instance. That mesh-wide Prometheus can either live outside of the mesh (external), or in one
> of the clusters within the mesh.

Reviewer 2: ```suggestion
The recommendations in this guide build on Prometheus
[hierarchical federation](https://prometheus.io/docs/prometheus/latest/federation/#hierarchical-federation).
Prometheus instances deployed locally to each cluster by Istio act as collectors that aggregate data hierarchically up to a production, mesh-wide Prometheus instance. The mesh-wide instance can exist in or outside the mesh.
```

## Multicluster Istio setup

> There are a couple of [multicluster deployment models](/docs/ops/deployment/deployment-models/#multiple-clusters)
> supported by Istio. You can follow the [multicluster installation](/docs/setup/install/multicluster/) section to setup
> your multicluster Istio. For the purposes of this guide, any of those approaches will work, with the following
> caveat:

Reviewer 2: ```suggestion
Follow the [multicluster installation](/docs/setup/install/multicluster/) section to setup
your Istio clusters in one of the supported [multicluster deployment models](/docs/ops/deployment/deployment-models/#multiple-clusters). For the purposes of this guide, any of those approaches will work, with the following
caveat:
```

**Ensure that a cluster-local Istio Prometheus instance is installed in each cluster.**

> Individual Istio deployments of Prometheus in each cluster are required to form the basis of cross-cluster monitoring by
> way of federation to a production-ready instance of Prometheus that runs externally or in one of the clusters.

Reviewer 2: ```suggestion
Deploy Prometheus in each cluster to achieve federated, cross-cluster monitoring.
```

> For multicluster deployments that use the `remote` profile, you must add the following to the `istioctl manifest` command:

Reviewer 2: ```suggestion
If you plan to use the `remote` profile for the multicluster deployment, add the following to the `istioctl manifest` command:
```

> {{< text bash >}}
> $ istioctl manifest apply -f --set addonComponents.prometheus.enabled=true
> {{< /text >}}

Reviewer: ```suggestion
{{< text plain >}}
--set addonComponents.prometheus.enabled=true
{{< /text >}}
```

Validate that you have an instance of Prometheus running in each cluster:

{{< text bash >}}
$ kubectl -n istio-system get services prometheus
NAME         TYPE        CLUSTER-IP   EXTERNAL-IP   PORT(S)    AGE
prometheus   ClusterIP   10.8.4.109   <none>        9090/TCP   20h
{{< /text >}}

## Configure Prometheus federation

### External production Prometheus

> There are several reasons why you may want to have a Prometheus instance running outside of your Istio deployment.
> Perhaps you want long-term monitoring disjoint from the cluster being monitored. Perhaps you want to monitor multiple
> separate meshes in a single place. Or maybe you have other motivations. Whatever your reason is, you’ll need some special
> configurations to make it all work.

Reviewer 2: ```suggestion
There are several reasons to run a Prometheus instance outside of your Istio deployment, including:
* Long-term monitoring independent from the cluster being monitored. 
* Monitoring multiple meshes in a single place. 

To achieve this, you’ll need additional configuration.
```

{{< image width="80%"
    link="./external-production-prometheus.svg"
    alt="Architecture of external Production Prometheus for monitoring multicluster Istio."
    caption="External Production Prometheus for monitoring multicluster Istio"
    >}}

{{< warning >}}

> This guide does not cover securing access to Prometheus. It demonstrates providing connectivity to cluster-local Prometheus
> instances in a few simple ways. For production use cases, it is recommended to secure access to each Prometheus endpoint
> with HTTPS, as well as taking appropriate precautions such as using an internal load-balancer instead of a publicly-accessible
> endpoint and/or properly configuring firewall rules.

Reviewer 2: ```suggestion
This guide demonstrates connectivity to cluster-local Prometheus
instances, but does not address security considerations. For production use, secure access to each Prometheus endpoint
with HTTPS. In addition, take precautions, such as using an internal load-balancer instead of public
endpoints and the appropriate firewall rule configuration.
```

> *Note:* This guide does not cover securing access to Prometheus. It demonstrates providing connectivity to cluster-local Prometheus
> instances in a few simple ways. For production use cases, it is recommended to secure access to each Prometheus endpoint
> with HTTPS, as well as taking appropriate precautions such as using an internal load-balancer instead of a publicly-accessible
> endpoint and/or properly configuring firewall rules.

Reviewer: ```suggestion
{{< warning >}}
This guide does not cover securing access to Prometheus. It demonstrates providing connectivity to cluster-local Prometheus
instances in a few simple ways. For production use cases, it is recommended to secure access to each Prometheus endpoint
with HTTPS, as well as taking appropriate precautions such as using an internal load-balancer instead of a publicly-accessible
endpoint and/or properly configuring firewall rules.
{{< /warning >}}
```

Author: done

Istio provides a way to expose cluster services externally via [Gateways](/docs/reference/config/networking/gateway/).
You can configure an ingress gateway for the cluster-local Prometheus, providing external connectivity to the in-cluster
Prometheus endpoint.

> For each cluster, follow the appropriate instructions from the [Remotely Accessing Telemetry Addons](/docs/tasks/observability/gateways/#option-1-secure-access-https) task. And you

Reviewer: ```suggestion
For each cluster, follow the appropriate instructions from the [Remotely Accessing Telemetry Addons](/docs/tasks/observability/gateways/#option-1-secure-access-https) task. Also note that you
```

> Also note that you **SHOULD** establish secure (HTTPS) access.

Reviewer 3: the `Remotely Accessing Telemetry Addons` page changed while this doc was in development. we should update this to reflect the current recommendations from that page -- everything going through port `80` and using `nip.io` domains.

Author: @douglas-reid I read that updated doc, which `nip.io` is not recommended in production settings. I'm not sure how to add descriptions here. Since we already pointing out to the secure access https section for the user to follow? Might not need to update anything? But yeah, need your more specific suggestions for this part.

> Also note that you **SHOULD** establish secure (HTTPS) access.

Reviewer 2: ```suggestion
Also, it is important to establish secure (HTTPS) access.
```

> After that, you will need to configure your external Prometheus instance to access the cluster-local Prometheus instances.
> This can be achieved with a configuration like the following (replacing the gateway address and cluster name):

Reviewer 2: ```suggestion
Next, configure your external Prometheus instance to access the cluster-local Prometheus instances using a configuration like the following (replacing the gateway address and cluster name):
```

{{< text yaml >}}
scrape_configs:
  - job_name: 'federate-{{CLUSTER_NAME}}'
    scrape_interval: 15s

    honor_labels: true
    metrics_path: '/federate'

    params:
      'match[]':
        - '{job="pilot"}'
        - '{job="envoy-stats"}'

    static_configs:
      - targets:
        - '{{GATEWAY_IP_ADDR}}:15030'
        labels:
          cluster: {{CLUSTER_NAME}}
{{< /text >}}

Notes:

> * `CLUSTER_NAME` should be set to the same value which is used to create the cluster (set via `values.global.multiCluster.clusterName`).

Reviewer 2: ```suggestion
* `CLUSTER_NAME` should be set to the same value that you used to create the cluster (set via `values.global.multiCluster.clusterName`).
```

* No authentication to the Prometheus endpoint(s) is provided. This means that anyone can query your
cluster-local Prometheus instances. This may not be desirable.

* Without proper HTTPS configuration of the gateway, everything is being transported via plaintext. This may not be
desirable.

### Production Prometheus from one of the clusters

> If you desire to run a Prometheus in one of the clusters, establish connectivity from the production instance of

Reviewer: ```suggestion
If you prefer to run the production Prometheus in one of the clusters, you need to establish connectivity from it to the other cluster-local Prometheus instances in the mesh.
```

> ### Production Prometheus from one of the clusters
> 
> If you prefer to run the production Prometheus in one of the clusters, you need to establish connectivity from it to
> the other cluster-local Prometheus instances in the mesh.

Reviewer 2: ```suggestion
### Run Prometheus on an internal cluster

If you prefer to run the production Prometheus in one of the clusters, establish connectivity from it to
the other cluster-local Prometheus instances in the mesh.
```

Author: @adammil2000 Would it be better to change this to `on an in-mesh cluster`?

Reviewer 2: Agreed, thanks!

> This is really just a customization of the process for external federation, which you can achieve by configuring the `Gateway`,
> `VirtualService`, and `DestinationRule` in each remote cluster.

Reviewer 2: ```suggestion
This example customizes the process for external federation by configuring the `Gateway`,
`VirtualService`, and `DestinationRule` in each remote cluster.
```

> 
> This is really just a customization of the process for external federation, which you can achieve by configuring the `Gateway`,
> `VirtualService`, and `DestinationRule` in each remote cluster.

Reviewer: ```suggestion
This is really just a variation of the configuration for external federation. In this case the configuration on the cluster running the production Prometheus is different from the configuration for remote cluster Prometheus scraping.
```

Author: I'd prefer the previous description. I think this change is a little bit ambiguous in terms of the remote cluster setup.

Reviewer: Sorry if my attempt at clarifying is not correct but the current wording is very unclear. I couldn't really understand what it's trying to say ... hence my best guess was wrong :)

1. `customization of the process` seems like a pretty awkward way to describe this. It's not really a `customization`, and there is no `process`. It's just a slightly different configuration model.
2. what's different about the `Gateway` configuration? You need it for both models.
3. What `VirtualService` and `DestinationRule` are you referring to? There is no mention of them in this doc. Either explain what needs to be configured, or don't mention them.

Author: yeah, agree with your above assertions. let me fix it.

{{< image width="80%"
    link="./in-mesh-production-prometheus.svg"
    alt="Architecture of in-mesh Production Prometheus for monitoring multicluster Istio."
    caption="In-mesh Production Prometheus for monitoring multicluster Istio"
    >}}

> Configure your production Prometheus to access both of the *local* and the *remote* Prometheus instances. This can be achieved
> by adding a configuration like the following for the *remote* clusters (replacing the service name and cluster name for each cluster):

Reviewer 2: ```suggestion
Configure your production Prometheus to access the local and remote Prometheus instances by adding a configuration for the remote clusters as shown (replace the service name and cluster name for each cluster):
```

{{< text yaml >}}
scrape_configs:
  - job_name: 'federate-{{CLUSTER_NAME}}'
    scrape_interval: 15s

    honor_labels: true
    metrics_path: '/federate'

    params:
      'match[]':
        - '{job="pilot"}'
        - '{job="envoy-stats"}'

    static_configs:
      - targets:
        - '{{GATEWAY_IP_ADDR}}:15030'
        labels:
          cluster: {{CLUSTER_NAME}}
{{< /text >}}

Then add a configuration like the following for the *local* cluster:

{{< text yaml >}}

> - job_name: 'federate-local'

Reviewer: Why is ther no `scrape_config:` and proper formatting (indent and blank lines) in this example, like the previous ones?

Author: - since I think according to the context, we are assuming to add this job to the above scrape_config? If we add scrape_config here, it would be causing confusion I guess. 
- Yes, indentation needed. Let me fix it!

Reviewer: If so, then that is totally unclear from the above sentence where it just says `add a configuration like the following ...`.

You would need to say something like `append the following config to the above scrape_config and then apply it in the *local* cluster:`. Or, you could show the full config with both, which would be easier for the user to cut and paste. Either way, my formatting question still applies. Why is it not indented 2 spaces, so it's correct when the user cut/pastes it? Also why not include the blank lines, for readability?

As an aside, most docs give the full command that the user can cut and paste, instead of just the YAML. Why are you not doing that here?

Author: Then, I think combine the two jobs to `scrape_config` to provide the users with a fully yaml would be the idea way to follow other docs' conventions. Let me gather them together.

the following (replacing the gateway address and cluster name):

{{< text yaml >}}
scrape_configs:
  - job_name: 'federate-{{CLUSTER_NAME}}'
    scrape_interval: 15s

    honor_labels: true
    metrics_path: '/federate'

    params:
      'match[]':
        - '{job="pilot"}'
        - '{job="envoy-stats"}'

    static_configs:
      - targets:
        - '{{GATEWAY_IP_ADDR}}:15030'
        labels:
          cluster: {{CLUSTER_NAME}}
{{< /text >}}

Notes:

* `CLUSTER_NAME` should be set to the same value which is used to create the cluster (set via `values.global.multiCluster.clusterName`).

* No Authentication to the Prometheus endpoint(s) is provided. This means that anyone will be able to query your
cluster-local Prometheus(es). This may not be desirable.

* Without proper HTTPS configuration of the gateway, everything is being transported via plaintext. This may not be
desirable.

### Production Prometheus from one of the clusters

If you desire to run a Prometheus in one of the clusters, you will need to establish connectivity from the production
instance of Prometheus to each of the cluster-local Prometheus(es) within the mesh.

> This is really just a customization of the process for external federation. You will need to establish the `Gateway`,
> `VirtualService`, and `DestinationRule` in each remote cluster as above in External Production Prometheus.

Reviewer 4: you can change this if you like, im not set on it
```suggestion
This is really just a customization of the process for external federation, which you can achieve by configuring the `Gateway`,
`VirtualService`, and `DestinationRule` in each remote cluster.
```

{{< image width="80%"
    link="./in-mesh-production-prometheus.svg"
    alt="Architecture of in-mesh Production Prometheus for monitoring multicluster Istio."
    caption="In-mesh Production Prometheus for monitoring multicluster Istio"
    >}}

> You need to configure your production Prometheus to access both of the *local* and the *remote* Prometheus(es). This can be achieved
> with config like the following for the *remote* clusters (replacing the service name and cluster name for each cluster):

Reviewer 4: ```suggestion
Configure your production Prometheus to access both the *local* and the *remote* Prometheus instances. This can be achieved
by adding a configuration like the following for the *remote* clusters (replacing the service name and cluster name for each cluster):
```

{{< text yaml >}}
scrape_configs:
  - job_name: 'federate-{{CLUSTER_NAME}}'
    scrape_interval: 15s

    honor_labels: true
    metrics_path: '/federate'

    params:
      'match[]':
        - '{job="pilot"}'
        - '{job="envoy-stats"}'

    static_configs:
      - targets:
        - '{{GATEWAY_IP_ADDR}}:15030'
        labels:
          cluster: {{CLUSTER_NAME}}
{{< /text >}}

> And add one configuration for the *local* cluster:

Reviewer 4: ```suggestion
Then add a configuration like the following for the *local* cluster:
```
